dll/win32/lpk/lpk.c

```diff
@@ -81,16 +78,30 @@ LpkExtTextOut(
     /* Check if the string requires complex script processing and not a "glyph indices" array */
     if (ScriptIsComplex(lpString, uCount, SIC_COMPLEX) == S_OK && !(fuOptions & ETO_GLYPH_INDEX))
     {
+        LPWORD glyphs = NULL;
+        LPWSTR reordered_str = HeapAlloc(GetProcessHeap(), 0, uCount * sizeof(WCHAR));
+        INT cGlyphs;
+        BOOL bResult;
+
         BIDI_Reorder(hdc, lpString, uCount, GCP_REORDER,
                      (fuOptions & ETO_RTLREADING) ? WINE_GCPW_FORCE_RTL : WINE_GCPW_FORCE_LTR,
-                     NULL, uCount, NULL, &glyphs, &cGlyphs);
+                     reordered_str, uCount, NULL, &glyphs, &cGlyphs);
 
-        fuOptions |= ETO_GLYPH_INDEX;
+        if (glyphs)
+        {
+            fuOptions |= ETO_GLYPH_INDEX;
 
-        if (uCount > cGlyphs)
-            cGlyphs = uCount;
+            if (uCount != cGlyphs)
+                uCount = cGlyphs;
+        }
 
-        return ExtTextOutW(hdc, x, y, fuOptions, lprc, (LPWSTR)glyphs, cGlyphs, lpDx);
+        bResult = ExtTextOutW(hdc, x, y, fuOptions, lprc,
+                              glyphs ? (LPWSTR)glyphs : reordered_str, uCount, lpDx);
```

Comment:
if `BIDI_Reorder` fails, is reordered_str reliable?

Reply:
First thing BIDI_Reorder does with reordered_str is to copy lpString to it

Reply:
That's not true, there's a HeapAlloc call before that happens, and that call can fail.

Reply:
https://github.com/wine-mirror/wine/blob/b6c7b2d322e0cd61eaeaac0d936dfc78c850f61b/dlls/gdi32/font.c#L2247
Wine doesn't checks if that allocation fails.
```
if (glyphs || reordered_str)
    //ExtTextOut with bidi strings
else
    //ExtTextOut with original lpString
```
^ will that make do?
Most of the code changes are based on how wine handles both variables.